### PR TITLE
Arm: Add Neon implementation for Morph dilation

### PR DESCRIPTION
**Pull request by kjg0724**
Follow-up to the ARM SIMD porting stream from #713/#717/#718, as called out in #711 (Morph::dilation has no Neon implementation).

## Change
Adds an ARM Neon port of `Morph::dilation`. The scalar/x86 algorithm dilates a "strong pixel" mask: for each pixel, if any of its 3x3 neighbours (including itself) equals a sentinel `Value`, the output becomes `Value`, otherwise it keeps the original value.

Since `extendBorderPel(1,1)` already writes real replicated border pixels on all sides, including corners, before the pixel loop (same as the scalar reference), the Neon kernel does plain unaligned loads for the left/right/top/bottom taps instead of the in-register lane-shift-and-carry approach the x86 SIMD version uses to avoid touching the border. This keeps the kernel considerably simpler than the x86 version while producing the same result as the scalar reference.

Processes 8 pixels per vector, with an overlapping tail strip for widths that are not a multiple of eight (safe since the destination buffer is disjoint from the source within an iteration) and a scalar fallback for widths under eight.

`Morph`'s constructor now takes an `enableOpt` flag (default true), mirroring the pattern already used for `FGAnalyzer` in #713/#717, so the unit test can construct a scalar-only reference instance alongside an optimized one. `FGAnalyzer` forwards its own `enableOpt` to its embedded `Morph` member.

## x86 fix (second commit)
This PR is the first unit test coverage `Morph::dilation` has ever had, and it surfaced a pre-existing bug in the x86 SIMD path (`dilation_SIMD` in `FGAX86.h`): the vectorized loop bound `x < width - 8` underflows for `width < 8` (width is unsigned), causing out-of-bounds SIMD loads, and the loop plus its tail-column write assume width is a multiple of eight and strictly greater than sixteen (at exactly 16 the AVX2 branch's own main loop runs zero times), so other widths either produce silently wrong output (width == 8 or 16) or write past the destination buffer (e.g. width == 44). Added a width check that delegates to the scalar reference (`dilation_core`) for widths that don't meet those assumptions, instead of duplicating its 3x3 test or touching the vectorized code.

## Performance (third commit)
The first version of the Neon kernel used a column-outer traversal with a rolling row mask to cut compare instructions, but that meant each 8-pixel-wide strip re-streamed the whole frame height, which falls out of cache well before 4K. Switched to row-outer traversal (fresh row-mask computations per output vector, no rolling state) instead.

Measured on this machine (Apple M-series, `-O3`):

| resolution | scalar | Neon |
|---|---|---|
| 1920x1080, iter=4 | 36.70 ms | 2.92 ms (12.6x) |
| 3840x2160, iter=4 | 99.17 ms | 12.11 ms (8.2x) |

Both resolutions scale consistently; the column-outer version measured during review dropped noticeably from 1080p to 4K for the same comparison.

## x86 fix, height == 1 (fourth commit)
The same width guard from the second commit missed another pre-existing edge case: the row-pointer setup in `dilation_SIMD` assumes at least two rows, so at `height == 1` the `y == 0` branch makes `p_buf_down` point one stride past the single row, an out-of-bounds read that produces silently wrong output. Extended the same guard to also fall back to `dilation_core` when `height == 1`.

## Testing
`MorphDilation` unit test compares scalar vs optimized output exactly across widths {4,8,16,24,32,44,64} (covering the scalar-fallback path, the exact-vector-width and AVX2-boundary sizes, the overlapping-tail path, and the width needed to reach the vectorized path at height 1) and heights {1,4,8,33,44,64} (including the height == 1 edge case from the fourth commit), with `numIter` in {1,2,3,4} and `Value` in {strongPel, 0}, matching the call-site patterns in `findMask`. Passes on both ARM and x86 (the x86 build exercises the fixes in the second and fourth commits), and the full unit test suite has no regressions on either.

### CODE REVIEW

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp [hunk 1]
```diff
@@ -96,6 +96,104 @@ void FGAnalyzer::_initFGAnalyzerARM()
...
+  return iter;
+}
+
+template <ARM_VEXT vext>
+void Morph::_initFGAMorphARM()
+{
+  dilation = dilationNeon;
+}
+
+template void Morph::_initFGAMorphARM<NEON>();
```
Comment (georges-arm) [inline source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: No need for a separate instantiation here, just specialise for `NEON` directly:

```
template<>
void Morph::_initFGAMorphARM<NEON>()
{
  ...
}
```

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp [hunk 2]
```diff
@@ -96,6 +96,104 @@ void FGAnalyzer::_initFGAnalyzerARM()
 
 template void FGAnalyzer::_initFGAnalyzerARM<NEON>();
```
Comment (georges-arm) [inline source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Move the other `init*` functions down to the bottom of the file so they are together? No need for a separate template instantiation here either.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp [hunk 3]
```diff
@@ -96,6 +96,104 @@ void FGAnalyzer::_initFGAnalyzerARM()
...
+          uint16x8_t strong  = vorrq_u16( mTop, vorrq_u16( mMid, mBot ) );
+          int16x8_t  res     = vbslq_s16( strong, vVal, cMid );
+          vst1q_s16( dst + (ptrdiff_t)y * dStr + x, res );
+
+          mTop = mMid;
+          mMid = mBot;
+          cMid = vld1q_s16( p + ( (ptrdiff_t)y + 1 ) * sStr );
+        }
+
+        if( x + 8 >= width ) break;
```
Comment (georges-arm) [inline source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Could this just be `x < width` in the loop condition?

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp [hunk 4]
```diff
@@ -96,6 +96,104 @@ void FGAnalyzer::_initFGAnalyzerARM()
...
+        for( int y = 0; y < height; y++ )
+        {
+          uint16x8_t mBot    = rowMask( p + ( (ptrdiff_t)y + 1 ) * sStr );
+          uint16x8_t strong  = vorrq_u16( mTop, vorrq_u16( mMid, mBot ) );
+          int16x8_t  res     = vbslq_s16( strong, vVal, cMid );
+          vst1q_s16( dst + (ptrdiff_t)y * dStr + x, res );
+
+          mTop = mMid;
+          mMid = mBot;
+          cMid = vld1q_s16( p + ( (ptrdiff_t)y + 1 ) * sStr );
```
Comment (georges-arm) [inline source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Is there a need for this to be loop-carried? Seems like it would be cleaner code to just load it as `cMid = vld1q_s16( p + (ptrdiff_t)y * sStr );` at the start of each `y` iteration?

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp [hunk 5]
```diff
@@ -96,6 +96,104 @@ void FGAnalyzer::_initFGAnalyzerARM()
...
+    const Pel*      src    = buff->get( compID ).buf;
+    const ptrdiff_t sStr   = buff->get( compID ).stride;
+    Pel*            dst    = Wbuf->bufs[0].buf;
+    const ptrdiff_t dStr   = Wbuf->bufs[0].stride;
+
+    if( width >= 8 )
+    {
+      const int16x8_t vVal = vdupq_n_s16( Value );
+
+      auto rowMask = [vVal]( const Pel* row ) -> uint16x8_t {
```
Comment (georges-arm) [inline source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: I think there is potentially a faster expression to do this since currently the compare results are mostly redundant. In particular two of the compare result vectors could be reused across `y` loop iterations I think so it may be worth inlining this and seeing what you can simplify.

At the very least you can avoid the `c` load with something like:

```
        int16x8_t  l = vld1q_s16( row - 1 );
        int16x8_t  r = vld1q_s16( row + 1 );
        int16x8_t  lm = vceqq_s16( l, vVal );
        int16x8_t  rm = vceqq_s16( r, vVal );
        uint16x8_t m = vqtbl2q_u8( {lm, rm}, /*some indices to extract c from lm/rm*/ );
        return vorrq_u16( m, vorrq_u16( lm, rm ) );
```

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 1 above)
Comment (kjg0724) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Done, both init functions now use explicit specialisation.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 2 above)
Comment (kjg0724) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Done, moved both to the end of the file next to each other.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 3 above)
Comment (kjg0724) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Yes, done, that's cleaner. Verified by hand that clamping x to width-8 before the loop body still lets the plain x<width condition terminate at the right point on the next increment.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 4 above)
Comment (kjg0724) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: This whole strip-outer traversal is gone now. Benchmarked it against a row-outer version and the strip-outer one loses noticeably at 4K, since each strip re-streams the full frame height. Switched to row-outer, so there's no loop-carried state across y at all anymore.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 5 above)
Comment (kjg0724) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Applied the c-load part: the centre row now computes its mask inline and reuses that load as the blend source instead of going through rowMask() twice for the same pixels.

Didn't do the cross-y reuse though. That needs consecutive y iterations for the same strip to be adjacent, which is exactly the traversal order I moved away from in the last commit (strip-outer was costing real throughput at 4K from re-streaming the frame per strip). Keeping reuse under row-outer would mean caching masks per strip across rows, which is more state for a saving that didn't show up in the benchmark even at 1080p where the frame fits in cache. Happy to revisit if you think it's worth it anyway.

**Comment by kjg0724** [comment]
Found one more edge case while extending test coverage: `dilation_SIMD` also has an out-of-bounds read at `height == 1`, since the `y == 0` branch makes `p_buf_down` point one stride past the single row. Same fallback pattern as the width guard above. Pushed in 9b9f4c2, with height 1 and width 32 added to the unit test to cover it.

File: source/Lib/CommonLib/arm/neon/FGA_neon.cpp (hunk 5 above)
Comment (georges-arm) [reply source/Lib/CommonLib/arm/neon/FGA_neon.cpp]: Thanks for the explanation, this seems fine now.

**Review by georges-arm — COMMENTED** [review]
Neon code seems fine now. There might be more tricks we can play by unrolling and using `EXT` to e.g. construct the cmp result for `p + 1` by using `EXT` to mix `p` and `p + 8` (from the next iteration of the loop) or `p - 1` by mixing `p - 8` (from the previous iteration of the loop) and `p`, but that is probably awkward to code and not guaranteed to be faster.

**Comment by kjg0724** [comment]
Good idea in principle, but the buffer only has a 2-pixel margin (`margin = m_convWidthG / 2` in `SEIFilmGrainAnalyzer.cpp`), so an EXT-based carry can't pull a full 8-lane lookback/lookahead at the row edges without reading out of bounds. The first and last iteration of each row would still need the plain-load path, same as how the x86 version special-cases its own x==0 boundary. And since NEON's unaligned loads don't carry a real throughput penalty to begin with, the actual win here is just trading load-port pressure for permute-port pressure, which is genuinely uncertain without measuring on the specific core. I'll leave it as a follow-up rather than fold it into this PR, but happy to revisit if it turns out to matter somewhere.

**Comment by adamjw24** [comment]
Thanks for providing the patch and the review!
